core/src/main/java/org/apache/iceberg/IncrementalDataTableScan.java

```diff
@@ -137,4 +136,29 @@ protected TableScan newRefinedScan(
     }
     return snapshots;
   }
+
+  private void validateSnapshotIdsRefinement(long newFromSnapshotId, long newToSnapshotId) {
+    Set<Long> snapshotIdsRange = Sets.newHashSet(
+        SnapshotUtil.snapshotIdsBetween(table(), fromSnapshotId, toSnapshotId));
+    // since snapshotIdsBetween return ids in range (fromSnapshotId, toSnapshotId]
+    snapshotIdsRange.add(fromSnapshotId);
+    Preconditions.checkArgument(
+        snapshotIdsRange.contains(newFromSnapshotId),
+        "newFromSnapshotId %s not in existing snapshot ids range (%s, %s]",
+        newFromSnapshotId, fromSnapshotId, newToSnapshotId);
+    Preconditions.checkArgument(
+        snapshotIdsRange.contains(newToSnapshotId),
+        "newToSnapshotId %s not in existing snapshot ids range (%s, %s]",
+        newToSnapshotId, fromSnapshotId, toSnapshotId);
+  }
+
+  private static void validateSnapshotIds(Table table, long fromSnapshotId, long toSnapshotId) {
+    Preconditions.checkArgument(fromSnapshotId != toSnapshotId, "fromSnapshotId and toSnapshotId cannot be the same");
```

Comment:
I'd like to avoid using argument names in error messages. Argument names aren't very readable for end users because they are dense and this practice requires changing strings and test cases if arguments get renamed later.

How about "From and to snapshot IDs cannot be the same: (%s, %s]", "From snapshot does not exist: %s", etc.

Reply:
Makes sense. Fixed

Reply:
Looks like you missed the test assertions.

Reply:
oops. I'll fix it